WalletWasabi/Crypto/Strobe.cs

```diff
@@ -0,0 +1,342 @@
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace WalletWasabi.Crypto
+{
+	// https://strobe.sourceforge.io/papers/strobe-20170130.pdf
+	// Based on Merlin framework small implementation: https://doc-internal.dalek.rs/src/merlin/strobe.rs.html
+	// https://github.com/dalek-cryptography/merlin/blob/1ed350bbc1d65f0a0697e0c20c48e11ec172c6ff/src/strobe.rs
+	public class Strobe128
+	{
+		private static readonly byte DDATA = 0x04;
+		private static readonly byte DRATE = 0x80;
+
+		private readonly byte[] _state = new byte[25 * 8]; // this is the block size used by keccak-f1600.
+		private byte _position = 0;
+		private byte _beginPosition = 0;
+		private StrobeFlags _currentFlags = 0;
+
+		// Let ˆr=r/8−2. This is the portion of the rate which is used for user data, measured in bytes.
+		private static readonly byte SpongeRate = 166;
+
+		public Strobe128(string procotol)
+		{
+			var V = "STROBEv1.0.2";
+			var S0 = ByteHelpers.Combine(
+				new byte[] { 1, (byte)(SpongeRate + 2), 1, 0, 1, 12 * 8 },  // F([[1, r/8, 1, 0, 1, 12·8]]
+				Encoding.UTF8.GetBytes(V));
+			Buffer.BlockCopy(S0, 0, _state, 0, S0.Length);
+			KeccakF1600(_state);
+			AddAssociatedMetaData(Encoding.UTF8.GetBytes(procotol), false);
+		}
+
+		private Strobe128(byte[] state, StrobeFlags flags, byte beginPosition, byte position)
+		{
+			Buffer.BlockCopy(state, 0, _state, 0, _state.Length);
+			_currentFlags = flags;
+			_beginPosition = beginPosition;
+			_position = position;
+		}
+
+		~Strobe128()
+		{
+			Array.Clear(_state, 0, _state.Length);
+		}
+
+		public void AddAssociatedMetaData(byte[] data, bool more)
+		{
+			BeginOperation(StrobeFlags.M | StrobeFlags.A, more);
+			Absorb(data);
+		}
+
+		public void AddAssociatedData(byte[] data, bool more)
+		{
+			BeginOperation(StrobeFlags.A, more);
+			Absorb(data);
+		}
+
+		public byte[] Prf(int count, bool more)
+		{
+			BeginOperation(StrobeFlags.I | StrobeFlags.A | StrobeFlags.C, more);
+			return Squeeze(new byte[count]);
+		}
+
+		public void Key(byte[] data, bool more)
+		{
+			BeginOperation(StrobeFlags.A | StrobeFlags.C, more);
+			Override(data);
+		}
+
+		public Strobe128 MakeCopy()
+		{
+			return new Strobe128(_state, _currentFlags, _beginPosition, _position);
+		}
+
+		public override string ToString()
```

Comment:
i think this should not be part of the public API, since the internal Keccak state should not leak. it's used in the tests, but should never be used outside of the tests.

In the Merlin implementation it's deliberately suppressed even from debug output: https://github.com/dalek-cryptography/merlin/blob/1ed350bbc1d65f0a0697e0c20c48e11ec172c6ff/src/strobe.rs#L41-L46

i think that's a good call, because if this information accidentally makes it to stderr or something and users inadvertently paste that then this may lead to loss of privacy or worse (although i'm having trouble coming up with a scenario)

Reply:
This should definitely not be part. I put it for testing only because there is not way to assert the internal state using the test vectors.

Reply:
it must be `public` to be used in the test, right? or does C# have a notion of "friend" classes? if not, maybe we should call it something ugly like `_internalToStringDoNotUse`

Reply:
I cannot have many options here. I renamed it and put it as internal.